src/GraphQL/DataLoader/DataLoaderBase.cs

```diff
@@ -1,52 +1,164 @@
+        /// <param name="equalityComparer">Specifies the equality comparer to be used, or null for the default equality comparer</param>
+        /// <param name="maxBatchSize">The maximum number of keys passed to the fetch function at a time</param>
+        public DataLoaderBase(bool caching, IEqualityComparer<TKey> equalityComparer, int maxBatchSize)
+        {
+            if (maxBatchSize < 1)
+                throw new ArgumentOutOfRangeException(nameof(maxBatchSize));
+            MaxBatchSize = maxBatchSize;
+            EqualityComparer = equalityComparer ?? EqualityComparer<TKey>.Default;
+            if (caching)
+                _cachedList = new Dictionary<TKey, DataLoaderPair<TKey, T>>(equalityComparer);
+        }
+
+        /// <summary>
+        /// Asynchronously load data for the provided given key
+        /// </summary>
+        /// <param name="key">Key to use for loading data</param>
+        /// <returns>
+        /// An object representing a pending operation
+        /// </returns>
+        public virtual IDataLoaderResult<T> LoadAsync(TKey key)
         {
-            if (!IsFetchNeeded())
+            lock (_sync)
             {
-                return;
+                //once it enters the lock, it is guaranteed to exit the lock, as it does not depend on external code
+                if (_cachedList != null)
+                {
+                    if (_cachedList.TryGetValue(key, out var ret2))
+                        return ret2;
+                }
+                if (_list != null)
+                {
+                    if (_list.TryGetValue(key, out var ret2))
+                        return ret2;
+
+                    if (_list.Count >= MaxBatchSize)
+                        _list = new DataLoaderList(this);
+                }
+                else
+                {
+                    _list = new DataLoaderList(this);
+                }
+                var ret = new DataLoaderPair<TKey, T>(_list, key);
+                _list.Add(key, ret);
+                _cachedList?.Add(key, ret);
+                return ret;
             }
+        }
 
-            var tcs = Interlocked.Exchange(ref _completionSource, CreateCompletionsSource());
+        /// <summary>
+        /// Returns an IDataLoaderResult that will return the specified value
+        /// </summary>
+        public static IDataLoaderResult<T> FromResult(T result) => new DataLoaderResult<T>(result);
 
-            if (cancellationToken.IsCancellationRequested)
-            {
-                // If cancellation has been requested already,
-                // set the task to cancelled without calling FetchAsync()
-                tcs.TrySetCanceled();
-                return;
-            }
+        /// <summary>
+        /// Returns an IDataLoaderResult that will return the specified asynchronous task
+        /// </summary>
+        public static IDataLoaderResult<T> FromResult(Task<T> result) => new DataLoaderResult<T>(result);
```

Comment:
I think these two static `FromResult` methods should be pulled out into their own extension methods class.

Reply:
I agree; I've moved them to the `DataLoaderExtensions` class (which already exists).

Reply:
I need to think a little more about this, actually....

Reply:
If they are an extension method, I think they need to be named something like `.ToDataLoaderResult()`.  But why should there be an extension method added to every object type, when this is only needed by a custom data loader class?

Originally I was probably imagining something like `Task.FromResult(x)`, and added it to `DataLoaderBase` for derived classes to use (e.g. `return FromResult(x);`).  But if we move it to `DataLoaderResult`, well there's already constructors within `DataLoaderResult` for that purpose - `new DataLoaderResult(x)` - which makes more sense than `DataLoaderResult.FromResult(x)`.

So I'm deleting those two methods.  They are not used by anything, do not change any tests, and are not mentioned in the docs.  They just aren't necessary.  Let me know if you have any comments @joemcbride @sungam3r

Reply:
Agree